akka-testkit-typed/src/main/scala/akka/testkit/typed/internal/TestKitUtils.scala

```diff
@@ -0,0 +1,71 @@
+/**
+ * Copyright (C) 2009-2018 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.testkit.typed.internal
+
+import akka.actor.typed.scaladsl.Behaviors
+import akka.actor.typed.{ ActorRef, ActorSystem, Behavior, Props }
+import akka.annotation.InternalApi
+
+import scala.concurrent.{ Await, TimeoutException }
+import scala.concurrent.duration.Duration
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] object ActorTestKitGuardian {
+  sealed trait TestKitCommand
+  final case class SpawnActor[T](name: String, behavior: Behavior[T], replyTo: ActorRef[ActorRef[T]], props: Props) extends TestKitCommand
+  final case class SpawnActorAnonymous[T](behavior: Behavior[T], replyTo: ActorRef[ActorRef[T]], props: Props) extends TestKitCommand
+
+  val testKitGuardian: Behavior[TestKitCommand] = Behaviors.immutable[TestKitCommand] {
+    case (ctx, SpawnActor(name, behavior, reply, props)) ⇒
+      reply ! ctx.spawn(behavior, name, props)
+      Behaviors.same
+    case (ctx, SpawnActorAnonymous(behavior, reply, props)) ⇒
+      reply ! ctx.spawnAnonymous(behavior, props)
+      Behaviors.same
+  }
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] object TestKitUtils {
+
+  // common internal utility impls for Java and Scala
+  private val TestKitRegex = """akka\.testkit\.typed\.(?:javadsl|scaladsl)\.ActorTestKit(?:\$.*)?""".r
+
+  def testNameFromCallStack(classToStartFrom: Class[_]): String = {
```

Comment:
do we have tests for this? is it also used from java tests and does it work there?

Reply:
Yes, both tests from Scala direct and derived spec and Java (old impl didn't cut it for all those cases). Also, verified it with all tests in akka-actor-typed-tests